flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/utils/TypeConversionsTest.java

```diff
@@ -0,0 +1,123 @@
+import java.util.Map;
+
+import static org.apache.flink.table.api.DataTypes.FIELD;
+import static org.apache.flink.table.api.DataTypes.INT;
+import static org.apache.flink.table.api.DataTypes.ROW;
+import static org.apache.flink.table.api.DataTypes.STRING;
+import static org.apache.flink.table.api.DataTypes.TIMESTAMP;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Tests for {@link TypeConversions}.
+ */
+public class TypeConversionsTest {
+	@Test
+	public void testExpandRowType() {
+		DataType dataType = ROW(
+			FIELD("f0", INT()),
+			FIELD("f1", STRING()),
+			FIELD("f2", TIMESTAMP(5).bridgedTo(Timestamp.class)),
+			FIELD("f3", TIMESTAMP(3)));
+		TableSchema schema = TypeConversions.expandCompositeTypeToSchema(dataType);
+
+		assertThat(
+			schema,
+			equalTo(
+				TableSchema.builder()
+					.field("f0", INT())
+					.field("f1", STRING())
+					.field("f2", TIMESTAMP(5).bridgedTo(Timestamp.class))
+					.field("f3", TIMESTAMP(3).bridgedTo(LocalDateTime.class))
+					.build()));
+	}
+
+	@Test
+	public void testExpandLegacyCompositeType() {
+		DataType dataType = TypeConversions.fromLegacyInfoToDataType(new TupleTypeInfo<>(
+			Types.STRING,
+			Types.INT,
+			Types.SQL_TIMESTAMP));
+		TableSchema schema = TypeConversions.expandCompositeTypeToSchema(dataType);
+
+		assertThat(
+			schema,
+			equalTo(
+				TableSchema.builder()
+					.field("f0", STRING())
+					.field("f1", INT())
+					.field("f2", TIMESTAMP(3).bridgedTo(Timestamp.class))
+					.build()));
+	}
+
+	@Test
+	public void testExpandStructuredType() {
+		StructuredType logicalType = StructuredType.newBuilder(ObjectIdentifier.of("catalog", "database", "type"))
+			.attributes(Arrays.asList(
+				new StructuredType.StructuredAttribute("f0", DataTypes.INT().getLogicalType()),
+				new StructuredType.StructuredAttribute("f1", DataTypes.STRING().getLogicalType()),
+				new StructuredType.StructuredAttribute("f2", DataTypes.TIMESTAMP(5).getLogicalType()),
+				new StructuredType.StructuredAttribute("f3", DataTypes.TIMESTAMP(3).getLogicalType())
+			))
+			.build();
+
+		Map<String, DataType> dataTypes = new HashMap<>();
+		dataTypes.put("f0", DataTypes.INT());
+		dataTypes.put("f1", DataTypes.STRING());
+		dataTypes.put("f2", DataTypes.TIMESTAMP(5).bridgedTo(Timestamp.class));
+		dataTypes.put("f3", DataTypes.TIMESTAMP(3));
+		FieldsDataType dataType = new FieldsDataType(logicalType, dataTypes);
+
+		TableSchema schema = TypeConversions.expandCompositeTypeToSchema(dataType);
+
+		assertThat(
+			schema,
+			equalTo(
+				TableSchema.builder()
+					.field("f0", INT())
+					.field("f1", STRING())
+					.field("f2", TIMESTAMP(5).bridgedTo(Timestamp.class))
+					.field("f3", TIMESTAMP(3).bridgedTo(LocalDateTime.class))
+					.build()));
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void testExpandThrowExceptionOnAtomicType() {
+		TypeConversions.expandCompositeTypeToSchema(DataTypes.TIMESTAMP());
```

Comment:
In the past we supported this case as well. The schema would be a field named `f0`. See `org.apache.flink.table.api.TableSchema#fromTypeInfo`. We could also think about adding the conversion logic to `TableSchema`.

Reply:
I did not want to support it on purpose in this utility method. Expanding an atomic type depends heavily on the context. I added more explanation in the javadoc what are the reasons why we should not support this method for atomic types.